# Model Evaluation and Selection

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('../datasets/model_list_summary.csv')

In [4]:
#This is the dataframe that collected all the best scores for each model type

pd.set_option('display.max_colwidth', None)
df

data_cleaning lemmatized  train_acc  test_acc  train_bal_acc  test_bal_acc  \
0       minimal         no   0.959632  0.902055       0.956994      0.897909   
1       minimal         no   0.968840  0.902055       0.967164      0.898785   
2       minimal         no   0.937681  0.917148       0.937164      0.916373   
3          deep         no   0.956028  0.897133       0.953674      0.893344   
4          deep        yes   0.969186  0.904216       0.968182      0.902283   
5          deep        yes   0.974247  0.898482       0.973243      0.896592   
6          deep         no   0.911493  0.897470       0.910347      0.895530   
7          deep         no   0.927238  0.903204       0.925733      0.901447   
8          deep        yes   0.939609  0.912985       0.938581      0.911462   
9          deep        yes   0.942982  0.905902       0.942322      0.904932   

                                                                                                                                                       params  
0        {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.001}  
1        {'logisticregression__C': 0.001, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': None}  
2     {'countvectorizer__max_features': 6700, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.12}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
7     {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9      {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}

In [4]:
# I will now be comparing the TEST accuracy scores and TEST balanced accuracy scores. 

In [5]:
df.describe()

train_acc   test_acc  train_bal_acc  test_bal_acc
count  10.000000  10.000000      10.000000     10.000000
mean    0.948694   0.904065       0.947340      0.901866
std     0.020366   0.006546       0.020200      0.007288
min     0.911493   0.897133       0.910347      0.893344
25%     0.938163   0.899376       0.937518      0.896922
50%     0.949505   0.902630       0.947998      0.900116
75%     0.966538   0.905481       0.964621      0.904270
max     0.974247   0.917148       0.973243      0.916373

In [6]:
#Ranking all models

df.sort_values(by='test_bal_acc', ascending=False)

data_cleaning lemmatized  train_acc  test_acc  train_bal_acc  test_bal_acc  \
2       minimal         no   0.937681  0.917148       0.937164      0.916373   
8          deep        yes   0.939609  0.912985       0.938581      0.911462   
9          deep        yes   0.942982  0.905902       0.942322      0.904932   
4          deep        yes   0.969186  0.904216       0.968182      0.902283   
7          deep         no   0.927238  0.903204       0.925733      0.901447   
1       minimal         no   0.968840  0.902055       0.967164      0.898785   
0       minimal         no   0.959632  0.902055       0.956994      0.897909   
5          deep        yes   0.974247  0.898482       0.973243      0.896592   
6          deep         no   0.911493  0.897470       0.910347      0.895530   
3          deep         no   0.956028  0.897133       0.953674      0.893344   

                                                                                                                                                       params  
2     {'countvectorizer__max_features': 6700, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.12}  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9      {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
7     {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
1        {'logisticregression__C': 0.001, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': None}  
0        {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.001}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

**It looks like the Bayes model (index 2) with minimal cleaning and no lemmatization had the best test accuracy score (.917) and test balanced accuracy score (.916) among all models**

### Now looking at only models created after deep cleaning...

In [14]:
#Sorting by accuracy score of test data

df[df['data_cleaning']=='deep'].sort_values(by='test_acc', ascending=False)

data_cleaning lemmatized  train_acc  test_acc  train_bal_acc  test_bal_acc  \
8          deep        yes   0.939609  0.912985       0.938581      0.911462   
9          deep        yes   0.942982  0.905902       0.942322      0.904932   
4          deep        yes   0.969186  0.904216       0.968182      0.902283   
7          deep         no   0.927238  0.903204       0.925733      0.901447   
5          deep        yes   0.974247  0.898482       0.973243      0.896592   
6          deep         no   0.911493  0.897470       0.910347      0.895530   
3          deep         no   0.956028  0.897133       0.953674      0.893344   

                                                                                                                                                       params  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9      {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
7     {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

In [15]:
#Sorting by balanced accuracy score of test data

df[df['data_cleaning']=='deep'].sort_values(by='test_bal_acc', ascending=False)

data_cleaning lemmatized  train_acc  test_acc  train_bal_acc  test_bal_acc  \
8          deep        yes   0.939609  0.912985       0.938581      0.911462   
9          deep        yes   0.942982  0.905902       0.942322      0.904932   
4          deep        yes   0.969186  0.904216       0.968182      0.902283   
7          deep         no   0.927238  0.903204       0.925733      0.901447   
5          deep        yes   0.974247  0.898482       0.973243      0.896592   
6          deep         no   0.911493  0.897470       0.910347      0.895530   
3          deep         no   0.956028  0.897133       0.953674      0.893344   

                                                                                                                                                       params  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9      {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
7     {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

***Looks like the order of highest to lowest scores is same for both metrics! The best model uses CountVectorizer and Bayes with the following parameters:***

- CountVectorizer: 
    - Maximum Features: 5,000
    - Ngram Range: (1,1)
    - Removing English stop words
- Bayes Multinomial:
    - Alpha:0.07

I would like to explore the different models more (tfdif vs count, lemmatized vs non-lemmatized). and in the interest of time, I edited the summary file and added couple of columns (estimators, vectorizer). I will read in this file and do my exploration.

In [16]:
df_models = pd.read_csv('../datasets/model_list_summary-final-with-additions.csv')

In [23]:
df_models.head(2)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
0    logreg      count       minimal         no   0.959632  0.902055   
1    logreg      tfidf       minimal         no   0.968840  0.902055   

   train_bal_acc  test_bal_acc  \
0       0.956994      0.897909   
1       0.967164      0.898785   

                                                                                                                                                 params  
0  {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.001}  
1  {'logisticregression__C': 0.001, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': None}

In [21]:
df_models[ (df_models['data_cleaning']=='deep') & (df_models['estimator']=='logreg')].sort_values(by='test_bal_acc', ascending=False)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
4    logreg      count          deep        yes   0.969186  0.904216   
5    logreg      tfidf          deep        yes   0.974247  0.898482   
3    logreg      count          deep         no   0.956028  0.897133   

   train_bal_acc  test_bal_acc  \
4       0.968182      0.902283   
5       0.973243      0.896592   
3       0.953674      0.893344   

                                                                                                                                                       params  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

The table above indicates that with the deep cleaning, and using logistic regression as the estimator, the balanced acccuracy score for the test set is best with CountVectorizer with lemmatization, followed by TfidfVectorizer with lemmatization. Lemmatization gave better models than non-lemmatization, and CountVectorization is better than TfidfVectorizer. (Using balanced accuracy score on test data as the metric)

In [22]:
df_models[ (df_models['data_cleaning']=='deep') & (df_models['estimator']=='bayes')].sort_values(by='test_bal_acc', ascending=False)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
8     bayes      count          deep        yes   0.939609  0.912985   
9     bayes      tfidf          deep        yes   0.942982  0.905902   
7     bayes      tfidf          deep         no   0.927238  0.903204   
6     bayes      count          deep         no   0.911493  0.897470   

   train_bal_acc  test_bal_acc  \
8       0.938581      0.911462   
9       0.942322      0.904932   
7       0.925733      0.901447   
6       0.910347      0.895530   

                                                                                                                                                    params  
8  {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9   {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
7  {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6   {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}

Like with logistic regression, doing lemmatization with Bayes estimator also produces better model as indicated by the table above.  With lemmatization, CountVectorizer performed better than TfidfVectorizer, but withouth lemmatization, TfidfVectorizer performed better than CountVectorizer. (Using balanced accuracy score on test data as the metric)

In [29]:
df_models[ (df_models['data_cleaning']=='deep') & (df_models['vectorizer']=='tfidf')].sort_values(by='test_bal_acc', ascending=False)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
9     bayes      tfidf          deep        yes   0.942982  0.905902   
7     bayes      tfidf          deep         no   0.927238  0.903204   
5    logreg      tfidf          deep        yes   0.974247  0.898482   

   train_bal_acc  test_bal_acc  \
9       0.942322      0.904932   
7       0.925733      0.901447   
5       0.973243      0.896592   

                                                                                                                                                      params  
9     {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
7    {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
5  {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}

The table above indicates that TfidfVectorizer works best with Bayes estimator, regardless of whether data is lemmatized or not, than Logistic Regression

In [30]:
df_models[ (df_models['data_cleaning']=='deep') & (df_models['vectorizer']=='count')].sort_values(by='test_bal_acc', ascending=False)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
8     bayes      count          deep        yes   0.939609  0.912985   
4    logreg      count          deep        yes   0.969186  0.904216   
6     bayes      count          deep         no   0.911493  0.897470   
3    logreg      count          deep         no   0.956028  0.897133   

   train_bal_acc  test_bal_acc  \
8       0.938581      0.911462   
4       0.968182      0.902283   
6       0.910347      0.895530   
3       0.953674      0.893344   

                                                                                                                                                       params  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

The table above indicates that CountVectorizer works best when data is lemmatized. 

In [25]:
#Ranking all models again

df_models.sort_values(by='test_bal_acc', ascending=False)

estimator vectorizer data_cleaning lemmatized  train_acc  test_acc  \
2     bayes      count       minimal         no   0.937681  0.917148   
8     bayes      count          deep        yes   0.939609  0.912985   
9     bayes      tfidf          deep        yes   0.942982  0.905902   
4    logreg      count          deep        yes   0.969186  0.904216   
7     bayes      tfidf          deep         no   0.927238  0.903204   
1    logreg      tfidf       minimal         no   0.968840  0.902055   
0    logreg      count       minimal         no   0.959632  0.902055   
5    logreg      tfidf          deep        yes   0.974247  0.898482   
6     bayes      count          deep         no   0.911493  0.897470   
3    logreg      count          deep         no   0.956028  0.897133   

   train_bal_acc  test_bal_acc  \
2       0.937164      0.916373   
8       0.938581      0.911462   
9       0.942322      0.904932   
4       0.968182      0.902283   
7       0.925733      0.901447   
1       0.967164      0.898785   
0       0.956994      0.897909   
5       0.973243      0.896592   
6       0.910347      0.895530   
3       0.953674      0.893344   

                                                                                                                                                       params  
2     {'countvectorizer__max_features': 6700, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.12}  
8     {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.07}  
9      {'multinomialnb__alpha': 0.3, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
4        {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.003}  
7     {'multinomialnb__alpha': 0.05, 'tfidfvectorizer__max_features': 3300, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
1        {'logisticregression__C': 0.001, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': None}  
0        {'countvectorizer__max_features': 5000, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': None, 'logisticregression__C': 0.001}  
5   {'logisticregression__C': 0.003, 'tfidfvectorizer__max_features': 5000, 'tfidfvectorizer__ngram_range': (1, 1), 'tfidfvectorizer__stop_words': 'english'}  
6      {'countvectorizer__max_features': 3000, 'countvectorizer__ngram_range': (1, 2), 'countvectorizer__stop_words': 'english', 'multinomialnb__alpha': 0.3}  
3  {'countvectorizer__max_features': 4500, 'countvectorizer__ngram_range': (1, 1), 'countvectorizer__stop_words': 'english', 'logisticregression__C': 0.0008}

#### Conclusion:

Based on the exploration and modeling of the data, below are my findings and recommendations:

1. Use Bayes model. The Bayes models performed better and had a higher balanced accuracy score better than logistic regression models.
2. Be discerning in data cleaning. As we see in the model scores created in this project, more data cleaning does not necessarily mean better models.
3. Lemmatize. For both Bayes and Logistic Regression, lemmatization brought up the balanced accuracy score. 
4. Consider both CountVectorizer and TfidfVectorizer since they both vary in performance depending on the estimators, lemmatization, and other factors.